File: grade-school/GradeSchool.fs
module GradeSchool

let empty: Map<int, string list> = 
    Map.empty<int, string list>

let add (student: string) (grade: int) (school: Map<int, string list>): Map<int, string list> = 
    let newGrade = 
        match school |> Map.tryFind grade with
        | Some students -> List.append [student] students |> List.sort
        | None -> [student]
    school |> Map.add grade newGrade

let roster (school: Map<int, string list>): (int * string list) list = 
    school |> Map.toList |> List.sortBy fst

let grade (number: int) (school: Map<int, string list>): string list = 
    match school |> Map.tryFind number with
    | Some students -> students |> List.sort
    | None -> []

File: go/grade-school/grade_school.go
package school

import "sort"

type Grade struct {
	Number   int
	Students []string
}

type School struct {
	grades []Grade
}

func New() *School {
	return &School{}
}

func (s *School) Add(student string, grade int) {
	for i := 0; i < len(s.grades); i++ {
		if s.grades[i].Number == grade {
			s.grades[i].Students = append(s.grades[i].Students, student)
			sort.Slice(s.grades[i].Students, func(a, b int) bool {
				return s.grades[i].Students[a] < s.grades[i].Students[b]
			})
			return
		}
	}
	s.grades = append(s.grades, Grade{grade, []string{student}})
	sort.Slice(s.grades, func(a, b int) bool { return s.grades[a].Number < s.grades[b].Number })
}

func (s *School) Grade(grade int) []string {
	for _, g := range s.grades {
		if g.Number == grade {
			return g.Students
		}
	}
	return []string{}
}

func (s *School) Enrollment() []Grade {
	return s.grades
}
